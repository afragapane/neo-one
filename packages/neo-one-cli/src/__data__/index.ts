export * from './getClients';
export * from './getContracts';
