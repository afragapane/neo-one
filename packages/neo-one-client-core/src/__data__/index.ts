export * from './keys';
export * from './data';
