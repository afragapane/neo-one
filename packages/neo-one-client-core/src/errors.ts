import { CustomError } from '@neo-one/utils';
import { ContractParameterTypeJSON } from './contractParameter';
import { ContractParameter } from './types';

export class InvalidFormatError extends CustomError {
  public readonly code: string;

  public constructor(reason?: string) {
    super(`Invalid format${reason === undefined ? '.' : `: ${reason}`}`);
    this.code = 'INVALID_FORMAT';
  }
}

export class VerifyError extends CustomError {
  public readonly code: string;

  public constructor(reason: string) {
    super(`Verification failed: ${reason}`);
    this.code = 'VERIFY';
  }
}

export class UnsignedBlockError extends CustomError {
  public readonly code: string;

  public constructor() {
    super('Block script does not exist because it has not been signed.');
    this.code = 'UNSIGNED_BLOCK';
  }
}

export class TooManyPublicKeysError extends CustomError {
  public readonly code: string = 'TOO_MANY_PUBLIC_KEYS';
}

export class InvalidNumberOfKeysError extends CustomError {
  public readonly code: string = 'INVALID_NUMBER_OF_KEYS';
}

export class InvalidContractParameterError extends CustomError {
  public readonly parameter: ContractParameter;
  public readonly expected: ReadonlyArray<ContractParameterTypeJSON>;
  public readonly code: string;

  public constructor(parameter: ContractParameter, expected: ReadonlyArray<ContractParameterTypeJSON>) {
    super(`Expected one of ${JSON.stringify(expected)} ` + `ContractParameterTypes, found ${parameter.type}`);

    this.parameter = parameter;
    this.expected = expected;
    this.code = 'INVALID_CONTRACT_PARAMETER';
  }
}
