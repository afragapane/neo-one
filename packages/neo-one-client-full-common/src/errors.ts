import { makeErrorWithCode } from '@neo-one/utils';

// tslint:disable-next-line export-name
export const InvalidContractPropertyStateError = makeErrorWithCode(
  'INVALID_CONTRACT_PROPERTY_STATE',
  (contractParameterType: number) => `Expected contract parameter type, found: ${contractParameterType.toString(16)}`,
);
