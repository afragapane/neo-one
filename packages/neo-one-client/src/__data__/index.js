/* @flow */
export { default as keys } from './keys';
export { default as transactions } from './transactions';
