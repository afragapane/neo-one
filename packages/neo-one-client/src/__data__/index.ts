export * from './keys';
export * from './data';
export * from './transactions';
export * from './factory';
