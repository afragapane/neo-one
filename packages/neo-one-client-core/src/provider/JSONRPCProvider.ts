import { Monitor } from '@neo-one/monitor';
import { JSONRPCError, UnknownBlockError } from '../errors';

export interface JSONRPCRequest {
  readonly method: string;
  // tslint:disable-next-line no-any
  readonly params?: any;
  readonly watchTimeoutMS?: number;
}

// tslint:disable-next-line no-any
export type JSONRPCResponse = any;

/**
 * Enables browser clients. See comlink and WorkerManager.
 */
export interface JSONRPCProviderManager {
  // tslint:disable no-method-signature unified-signatures
  getInstance(): Promise<JSONRPCProvider>;
  // tslint:enable no-method-signature unified-signatures
}

export abstract class JSONRPCProvider {
  public abstract request(req: JSONRPCRequest, monitor?: Monitor): Promise<JSONRPCResponse>;
  // tslint:disable-next-line no-any
  protected readonly handleResponse = (responseJSON: any): any => {
    if (responseJSON.error !== undefined) {
      if (responseJSON.error.code === -100 && responseJSON.error.message === 'Unknown block') {
        throw new UnknownBlockError();
      }
      throw new JSONRPCError(responseJSON.error);
    }

    return responseJSON.result;
  };
}
