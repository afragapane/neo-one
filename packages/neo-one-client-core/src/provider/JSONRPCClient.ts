import {
  AddressString,
  addressToScriptHash,
  BlockJSON,
  BufferString,
  ConfirmedTransactionJSON,
  ContractJSON,
  ContractParameterJSON,
  GetOptions,
  Hash256String,
  HeaderJSON,
  InvocationResultJSON,
  Peer,
  PluginJSON,
  RelayTransactionResultJSON,
  StorageItemJSON,
  ValidatorJSON,
} from '@neo-one/client-common';
import BigNumber from 'bignumber.js';
import { RelayTransactionError } from '../errors';
import { JSONRPCProvider, JSONRPCProviderManager } from './JSONRPCProvider';

export class JSONRPCClient {
  private readonly provider: JSONRPCProvider | JSONRPCProviderManager;

  public constructor(provider: JSONRPCProvider | JSONRPCProviderManager) {
    this.provider = provider;
  }

  // Blockchain

  public async getBlock(hashOrIndex: Hash256String | number, options: GetOptions = {}): Promise<BlockJSON> {
    const { timeoutMS } = options;

    return this.withInstance(async (provider) =>
      provider.request({
        method: 'getblock',
        params: [hashOrIndex, 1],
        watchTimeoutMS: timeoutMS,
      }),
    );
  }

  public async getBestBlockHash(): Promise<string> {
    return this.withInstance(async (provider) => provider.request({ method: 'getbestblockhash' }));
  }

  public async getBlockCount(): Promise<number> {
    return this.withInstance(async (provider) => provider.request({ method: 'getblockcount' }));
  }

  public async getBlockHash(index: number): Promise<Hash256String> {
    return this.withInstance(async (provider) => provider.request({ method: 'getblockhash', params: [index] }));
  }

  public async getBlockHeader(hashOrIndex: Hash256String | number, options: GetOptions = {}): Promise<HeaderJSON> {
    const { timeoutMS } = options;

    return this.withInstance(async (provider) =>
      provider.request({
        method: 'getblockheader',
        params: [hashOrIndex, 1],
        watchTimeoutMS: timeoutMS,
      }),
    );
  }
  public async getBlockSysFee(index: number): Promise<BigNumber> {
    return this.withInstance(async (provider) =>
      provider.request({ method: 'getblocksysfee', params: [index] }).then((value) => new BigNumber(value)),
    );
  }

  public async getContract(address: AddressString): Promise<ContractJSON> {
    return this.withInstance(async (provider) =>
      provider.request({
        method: 'getcontractstate',
        params: [addressToScriptHash(address)],
      }),
    );
  }

  public async getMemPool(): Promise<readonly string[]> {
    return this.withInstance(async (provider) => provider.request({ method: 'getrawmempool' }));
  }

  public async getTransaction(hash: Hash256String, options: GetOptions = {}): Promise<ConfirmedTransactionJSON> {
    const { timeoutMS } = options;

    return this.withInstance(async (provider) =>
      provider.request({
        method: 'getrawtransaction',
        params: [hash, 1],
        watchTimeoutMS: timeoutMS,
      }),
    );
  }

  public async getTransactionHeight(hash: Hash256String): Promise<number> {
    return this.withInstance(async (provider) =>
      provider.request({
        method: 'gettransactionheight',
        params: [hash],
      }),
    );
  }

  public async getStorage(
    scriptHashOrId: AddressString | number,
    key: BufferString,
  ): Promise<readonly StorageItemJSON[]> {
    return this.withInstance(async (provider) =>
      provider.request({
        method: 'getstorage',
        params: [typeof scriptHashOrId === 'number' ? scriptHashOrId : addressToScriptHash(scriptHashOrId), key],
      }),
    );
  }

  public async getValidators(): Promise<ValidatorJSON> {
    return this.withInstance(async (provider) =>
      provider.request({
        method: 'getvalidators',
      }),
    );
  }

  // Node

  public async getConnectionCount(): Promise<number> {
    return this.withInstance(async (provider) =>
      provider.request({
        method: 'getconnectioncount',
      }),
    );
  }

  public async getConnectedPeers(): Promise<readonly Peer[]> {
    return this.withInstance(async (provider) =>
      provider
        .request({
          method: 'getpeers',
        })
        .then((result) => result.connected),
    );
  }

  public async getVersion() {
    return this.withInstance(async (provider) =>
      provider.request({
        method: 'getversion',
      }),
    );
  }

  public async sendTransaction(value: BufferString, options: GetOptions = {}): Promise<RelayTransactionResultJSON> {
    const { timeoutMS } = options;

    return this.withInstance(async (provider) =>
      provider
        .request({
          method: 'sendrawtransaction',
          params: [value],
          watchTimeoutMS: timeoutMS,
        })
        .catch((error) => {
          const [message, code]: [string, string] = error.message.split(':');
          if (error.code === 'JSON_RPC' && code === '-110') {
            throw new RelayTransactionError(message);
          }

          throw error;
        }),
    );
  }

  // Smart Contract

  public async invokeFunction(
    contract: AddressString,
    method: string,
    args: readonly ContractParameterJSON[],
  ): Promise<InvocationResultJSON> {
    return this.withInstance(async (provider) =>
      provider.request({
        method: 'invokefunction',
        params: [addressToScriptHash(contract), method, args],
      }),
    );
  }

  public async invokeScript(script: BufferString): Promise<InvocationResultJSON> {
    return this.withInstance(async (provider) =>
      provider.request({
        method: 'invokescript',
        params: [script],
      }),
    );
  }

  // Utilities

  public async listPlugins(): Promise<PluginJSON> {
    return this.withInstance(async (provider) =>
      provider.request({
        method: 'listplugins',
      }),
    );
  }

  // Dev

  // public async getNetworkSettings(): Promise<NetworkSettingsJSON> {
  //   return this.withInstance(async (provider) =>
  //     provider.request({
  //       method: 'getnetworksettings',
  //     }),
  //   );
  // }

  // public async runConsensusNow(): Promise<void> {
  //   return this.withInstance(async (provider) =>
  //     provider.request({
  //       method: 'runconsensusnow',
  //     }),
  //   );
  // }

  // public async updateSettings(options: Partial<PrivateNetworkSettings>): Promise<void> {
  //   return this.withInstance(async (provider) =>
  //     provider.request({
  //       method: 'updatesettings',
  //       params: [options],
  //     }),
  //   );
  // }

  // public async getSettings(): Promise<PrivateNetworkSettings> {
  //   return this.withInstance(async (provider) =>
  //     provider.request({
  //       method: 'getsettings',
  //     }),
  //   );
  // }

  // public async fastForwardOffset(seconds: number): Promise<void> {
  //   return this.withInstance(async (provider) =>
  //     provider.request({
  //       method: 'fastforwardoffset',
  //       params: [seconds],
  //     }),
  //   );
  // }

  // public async fastForwardToTime(seconds: number): Promise<void> {
  //   return this.withInstance(async (provider) =>
  //     provider.request({
  //       method: 'fastforwardtotime',
  //       params: [seconds],
  //     }),
  //   );
  // }

  // public async reset(): Promise<void> {
  //   return this.withInstance(async (provider) => provider.request({ method: 'reset' }));
  // }

  // public async getNEOTrackerURL(): Promise<string | undefined> {
  //   return this.withInstance(async (provider) => provider.request({ method: 'getneotrackerurl' }));
  // }

  // public async resetProject(): Promise<void> {
  //   return this.withInstance(async (provider) => provider.request({ method: 'resetproject' }));
  // }

  private async withInstance<TResult>(func: (instance: JSONRPCProvider) => Promise<TResult>): Promise<TResult>;
  private async withInstance<TResult>(func: (instance: JSONRPCProvider) => TResult): Promise<TResult> {
    if (this.provider instanceof JSONRPCProvider) {
      // tslint:disable-next-line no-any
      return func(this.provider as any);
    }

    const instance = await this.provider.getInstance();

    return func(instance);
  }
}
