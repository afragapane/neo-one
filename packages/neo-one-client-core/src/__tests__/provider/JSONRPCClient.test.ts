// tslint:disable no-object-mutation
import BigNumber from 'bignumber.js';
import { data, factory, keys } from '../../__data__';
import { JSONRPCError, RelayTransactionError } from '../../errors';
import { JSONRPCClient, JSONRPCProvider } from '../../provider';

const mockRequest = jest.fn();
class MockJSONRPCProvider extends JSONRPCProvider {
  // tslint:disable-next-line readonly-keyword
  public request = mockRequest;
}

describe('JSONRPCClient', () => {
  const provider = new MockJSONRPCProvider();
  const client = new JSONRPCClient(provider);

  beforeEach(() => {
    mockRequest.mockReset();
  });

  test('getBlock - hash', async () => {
    const value = factory.createBlockJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getBlock(data.hash256s.a);

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getBlock - index', async () => {
    const value = factory.createBlockJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getBlock(10, { timeoutMS: 1000 });

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getBlockHeader - hash', async () => {
    const value = factory.createHeaderJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getBlockHeader(data.hash256s.a);

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getBlockHeader - index', async () => {
    const value = factory.createHeaderJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getBlockHeader(10, { timeoutMS: 1000 });

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getBestBlockHash', async () => {
    const value = data.hash256s.a;
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getBestBlockHash();

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getBlockCount', async () => {
    const value = 10;
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getBlockCount();

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getBlockHash', async () => {
    const value = data.hash256s.a;
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getBlockHash(10);

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getBlockSysFee', async () => {
    const value = new BigNumber('10');
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getBlockSysFee(10);

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getContract', async () => {
    const value = factory.createContractJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getContract(keys[0].address);

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getMemPool', async () => {
    const value = [data.hash256s.a];
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getMemPool();

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getTransaction', async () => {
    const value = factory.createTransactionJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getTransaction(data.hash256s.a, { timeoutMS: 1000 });

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getTransactionHeight', async () => {
    const value = 1000;
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getTransactionHeight(data.hash256s.a);

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getStorage', async () => {
    const value = [factory.createStorageItemJSON()];
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getStorage(keys[0].address, data.buffers.b);

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getValidators', async () => {
    const value = [factory.createValidatorJSON()];
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getValidators();

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getConnectedPeers', async () => {
    const value = { connected: [factory.createPeerJSON()] };
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getConnectedPeers();

    expect(result).toEqual(value.connected);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getConnectionCount', async () => {
    const value = 9;
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getConnectionCount();

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('getVersion', async () => {
    const value = factory.createVersionJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.getVersion();

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('invokeScript', async () => {
    const value = factory.createInvocationResultSuccessJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.invokeScript(data.buffers.a);

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('invokeFunction', async () => {
    const value = factory.createInvocationResultErrorJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.invokeFunction(keys[0].address, 'balanceOf', [
      { name: 'param', type: 'Hash160', value: keys[1].address },
    ]);

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('sendTransaction', async () => {
    const value = factory.createTransactionJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.sendTransaction(data.buffers.a);

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('sendTransaction - relay error', async () => {
    const message = 'error';
    const rpcErr = new JSONRPCError({ message, code: -110 });
    mockRequest.mockImplementationOnce(async () => Promise.reject(rpcErr));

    const result = client.sendTransaction(data.buffers.a);

    await expect(result).rejects.toEqual(new RelayTransactionError(message));
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('sendTransaction - other error', async () => {
    const error = new Error('error');
    mockRequest.mockImplementationOnce(async () => Promise.reject(error));

    const result = client.sendTransaction(data.buffers.a);

    await expect(result).rejects.toEqual(error);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  test('listPlugins', async () => {
    const value = factory.createPluginJSON();
    mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

    const result = await client.listPlugins();

    expect(result).toEqual(value);
    expect(mockRequest.mock.calls).toMatchSnapshot();
  });

  // test('getNetworkSettings', async () => {
  //   const value = factory.createNetworkSettingsJSON();
  //   mockRequest.mockImplementationOnce(async () => Promise.resolve(value));

  //   const result = await client.getNetworkSettings();

  //   expect(result).toEqual(value);
  //   expect(mockRequest.mock.calls).toMatchSnapshot();
  // });

  // test('runConsensusNow', async () => {
  //   mockRequest.mockImplementationOnce(async () => Promise.resolve());

  //   await client.runConsensusNow();

  //   expect(mockRequest.mock.calls).toMatchSnapshot();
  // });

  // test('updateSettings', async () => {
  //   mockRequest.mockImplementationOnce(async () => Promise.resolve());

  //   await client.updateSettings({ secondsPerBlock: 10 });

  //   expect(mockRequest.mock.calls).toMatchSnapshot();
  // });

  // test('fastForwardOffset', async () => {
  //   mockRequest.mockImplementationOnce(async () => Promise.resolve());

  //   await client.fastForwardOffset(10);

  //   expect(mockRequest.mock.calls).toMatchSnapshot();
  // });

  // test('fastForwardToTime', async () => {
  //   mockRequest.mockImplementationOnce(async () => Promise.resolve());

  //   await client.fastForwardToTime(10);

  //   expect(mockRequest.mock.calls).toMatchSnapshot();
  // });

  // test('reset', async () => {
  //   mockRequest.mockImplementationOnce(async () => Promise.resolve());

  //   await client.reset();

  //   expect(mockRequest.mock.calls).toMatchSnapshot();
  // });
});
