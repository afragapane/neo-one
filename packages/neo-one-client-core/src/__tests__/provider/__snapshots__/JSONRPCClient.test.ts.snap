// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONRPCClient getBestBlockHash 1`] = `
Array [
  Array [
    Object {
      "method": "getbestblockhash",
    },
  ],
]
`;

exports[`JSONRPCClient getBlock - hash 1`] = `
Array [
  Array [
    Object {
      "method": "getblock",
      "params": Array [
        "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
        1,
      ],
      "watchTimeoutMS": undefined,
    },
  ],
]
`;

exports[`JSONRPCClient getBlock - index 1`] = `
Array [
  Array [
    Object {
      "method": "getblock",
      "params": Array [
        10,
        1,
      ],
      "watchTimeoutMS": 1000,
    },
  ],
]
`;

exports[`JSONRPCClient getBlockCount 1`] = `
Array [
  Array [
    Object {
      "method": "getblockcount",
    },
  ],
]
`;

exports[`JSONRPCClient getBlockHash 1`] = `
Array [
  Array [
    Object {
      "method": "getblockhash",
      "params": Array [
        10,
      ],
    },
  ],
]
`;

exports[`JSONRPCClient getBlockHeader - hash 1`] = `
Array [
  Array [
    Object {
      "method": "getblockheader",
      "params": Array [
        "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
        1,
      ],
      "watchTimeoutMS": undefined,
    },
  ],
]
`;

exports[`JSONRPCClient getBlockHeader - index 1`] = `
Array [
  Array [
    Object {
      "method": "getblockheader",
      "params": Array [
        10,
        1,
      ],
      "watchTimeoutMS": 1000,
    },
  ],
]
`;

exports[`JSONRPCClient getBlockSysFee 1`] = `
Array [
  Array [
    Object {
      "method": "getblocksysfee",
      "params": Array [
        10,
      ],
    },
  ],
]
`;

exports[`JSONRPCClient getConnectedPeers 1`] = `
Array [
  Array [
    Object {
      "method": "getpeers",
    },
  ],
]
`;

exports[`JSONRPCClient getConnectionCount 1`] = `
Array [
  Array [
    Object {
      "method": "getconnectioncount",
    },
  ],
]
`;

exports[`JSONRPCClient getContract 1`] = `
Array [
  Array [
    Object {
      "method": "getcontractstate",
      "params": Array [
        "0xcef0c0fdcfe7838eff6ff104f9cdec2922297537",
      ],
    },
  ],
]
`;

exports[`JSONRPCClient getMemPool 1`] = `
Array [
  Array [
    Object {
      "method": "getrawmempool",
    },
  ],
]
`;

exports[`JSONRPCClient getStorage 1`] = `
Array [
  Array [
    Object {
      "method": "getstorage",
      "params": Array [
        "0xcef0c0fdcfe7838eff6ff104f9cdec2922297537",
        "5f8d70",
      ],
    },
  ],
]
`;

exports[`JSONRPCClient getTransaction 1`] = `
Array [
  Array [
    Object {
      "method": "getrawtransaction",
      "params": Array [
        "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
        1,
      ],
      "watchTimeoutMS": 1000,
    },
  ],
]
`;

exports[`JSONRPCClient getTransactionHeight 1`] = `
Array [
  Array [
    Object {
      "method": "gettransactionheight",
      "params": Array [
        "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
      ],
    },
  ],
]
`;

exports[`JSONRPCClient getValidators 1`] = `
Array [
  Array [
    Object {
      "method": "getvalidators",
    },
  ],
]
`;

exports[`JSONRPCClient getVersion 1`] = `
Array [
  Array [
    Object {
      "method": "getversion",
    },
  ],
]
`;

exports[`JSONRPCClient invokeFunction 1`] = `
Array [
  Array [
    Object {
      "method": "invokefunction",
      "params": Array [
        "0xcef0c0fdcfe7838eff6ff104f9cdec2922297537",
        "balanceOf",
        Array [
          Object {
            "name": "param",
            "type": "Hash160",
            "value": "ALfnhLg7rUyL6Jr98bzzoxz5J7m64fbR4s",
          },
        ],
      ],
    },
  ],
]
`;

exports[`JSONRPCClient invokeScript 1`] = `
Array [
  Array [
    Object {
      "method": "invokescript",
      "params": Array [
        "b500",
      ],
    },
  ],
]
`;

exports[`JSONRPCClient listPlugins 1`] = `
Array [
  Array [
    Object {
      "method": "listplugins",
    },
  ],
]
`;

exports[`JSONRPCClient sendTransaction - other error 1`] = `
Array [
  Array [
    Object {
      "method": "sendrawtransaction",
      "params": Array [
        "b500",
      ],
      "watchTimeoutMS": undefined,
    },
  ],
]
`;

exports[`JSONRPCClient sendTransaction - relay error 1`] = `
Array [
  Array [
    Object {
      "method": "sendrawtransaction",
      "params": Array [
        "b500",
      ],
      "watchTimeoutMS": undefined,
    },
  ],
]
`;

exports[`JSONRPCClient sendTransaction 1`] = `
Array [
  Array [
    Object {
      "method": "sendrawtransaction",
      "params": Array [
        "b500",
      ],
      "watchTimeoutMS": undefined,
    },
  ],
]
`;
